### DISCUSSION TITLE
How to describe optional security?

### DISCUSSION THREAD

**Discussion by johnholiver** [goadesign/goa#2802] [Q&A] [answered]
This is a valid OpenAPI spec:
```
security:
      - {}
      - jwt_header_Authorization: []
```

It indicates that the endpoint can be accessed without security and with security (jwt in this case).
It has been a feature requested on the OpenAPI community years ago, as can be seen [here](https://github.com/OAI/OpenAPI-Specification/issues/14).

Is there a way to represent this in Goa? I don't see an object that can represent NoSecurity as a scheme on the DSL. Actually, there, NoSecurity() seems to be used to reset the need for security of an endpoint.

I'll try to make the parameter not required on the Payload, but:
1) I don't know if this will work
2) I see that as a problem while generating the OpenAPI

**Comment by raphael** [comment, marked as answer]
The code generation algorithm does [explicitly handle](https://github.com/goadesign/goa/blob/v3/http/codegen/server.go#L562) the case where the payload attribute used to map the credential is not required and in this case provides a nil value to the authorization function. So at least mechanically you should be able to implement the behavior you are after. It should be possible to change the OpenAPI spec generation so it matches the above by doing a similar check. [The current code](https://github.com/goadesign/goa/blob/e77e3fbe77e03c97e4abe31ecae5240987e9c18d/http/codegen/openapi/v3/builder.go#L429) that generates the OpenAPI security section does not have access to the payload so it would have to be modified so that it can access the method payload and check whether the credential field is required or not.
